### ISSUE TITLE
Wrong DateTimeIndex slicing when freq is None

### ISSUE THREAD

**Issue by michcio1234** [dask/dask#2389]
When Dask DataFrame is constructed from Pandas DataFrame with `DateTimeIndex` but `freq` is `None`, dask `loc` slicing fails.

An example:
```
import dask.dataframe as dd
import pandas as pd

# Construct some data with DateTimeIndex
datetime_index = pd.date_range('2016-01-01', '2016-01-31', freq='12h')
datetime_index.freq = None  # FORGET FREQUENCY
df = pd.DataFrame({'num': range(len(datetime_index))}, index=datetime_index)

ddf = dd.from_pandas(df, npartitions=1)

# Request a subset of data in two ways
buggy = ddf.loc['2016-01-03':'2016-01-05', :].compute()
ok = ddf.compute().loc['2016-01-03':'2016-01-05', :]

print('\n loc-index-compute')
print(buggy)
print('\n compute-loc-index')
print(ok)

```

Outcome:
```
 loc-index-compute
                     num
2016-01-04 00:00:00    6
2016-01-04 12:00:00    7
2016-01-05 00:00:00    8

 compute-loc-index
                     num
2016-01-03 00:00:00    4
2016-01-03 12:00:00    5
2016-01-04 00:00:00    6
2016-01-04 12:00:00    7
2016-01-05 00:00:00    8
2016-01-05 12:00:00    9

Process finished with exit code 0
```

What happens:
Slicing range is not determined properly. We wanted to get slice from 2016-01-03 00:00:00 to 2016-01-05 23:59:59. Instead, we got a slice from 2016-01-03 23:59:59 to 2016-01-05 00:00:00.

Diagnose:
This happens because of `dask.dataframe.utils.meta_nonempty()` function that calls  `dask.dataframe.utils._nonempty_index()`. What this function does is:
```
    elif typ is pd.DatetimeIndex:
        start = '1970-01-01'
        data = [start, start] if idx.freq is None else None
        return pd.DatetimeIndex(data, start=start, periods=2, freq=idx.freq,
                                tz=idx.tz, name=idx.name)
```
Then, in `pandas.core.indexes.datetimes._maybe_cast_slice_bound()` this index is considered monotonically decreasing, causing described behaviour.

Dask version: 0.14.3, installed from source (same happens when installed from PyPI)
Pandas version: 0.20.1 (same happens in 0.19.2)

**Comment by TomAugspurger** [comment]
Thanks for the (excellent) bug report. I have a potential fix ready, just making sure it doesn't break anything else.

**Comment by michcio1234** [comment]
Thank you for the quick fix. By the way - doesn't `dask.dataframe.utils._nonempty_series` function do the same thing? It creates a Series with 2 identical timestamps.

**Comment by TomAugspurger** [comment]
> By the way - doesn't dask.dataframe.utils._nonempty_series function do the same thing?

It does, though I don't think you can get in a state where that metadata is used for indexing.

```python
In [3]: a = pd.Series([pd.Timestamp('2017'), pd.Timestamp('2017-01-02'), pd.Timestamp('2017-01-03'), pd.Timestamp('2017-01-05')], name='a')

In [4]: b = pd.Series([pd.Timestamp('2017'), pd.Timestamp('2017-01-02'), pd.Timestamp('2017-01-03'), pd.Timestamp('2017-01-05')], name='b')

In [6]: import dask.dataframe as dd

In [7]: dd.concat([a, b], axis=1).set_index('a').index._meta_nonempty
Out[7]: DatetimeIndex(['1970-01-01', '1970-01-02'], dtype='datetime64[ns]', name='a', freq=None)
```

If you have an example otherwise, I'll change that as well, but I'd prefer to leave it as is otherwise.

FYI, this will hopefully be fixed in pandas 0.20.2, which will be released in a couple days.
